lms/djangoapps/bookmarks_api/views.py

```diff
@@ -0,0 +1,210 @@
+
+    paginate_by = 1000
+    paginate_by_param = 'page_size'
+    pagination_serializer_class = PaginationSerializer
+    serializer_class = serializers.BookmarkSerializer
+
+    def get_queryset(self):
+        course_id = self.request.QUERY_PARAMS.get('course_id', None)
+        fields = self.request.QUERY_PARAMS.get('fields', None)
+
+        if not course_id:
+            return Response(status=status.HTTP_400_BAD_REQUEST)
+        try:
+            course_key = CourseKey.from_string(course_id)
+        except InvalidKeyError:
+            log.error("Invalid course id '{course_id}'")
+            return list()
+
+        results_queryset = Bookmark.objects.filter(course_key=course_key, user__id=self.request.user.id).values("course_key", "usage_key", "display_name", "created").order_by('-created')
+        # serializer = serializers.BookmarkSerializer(results, many=True, fields_to_remove=['path'])
+
+        return results_queryset
+
+    def post(self, request):
+        """
+        Create a new bookmark.
+
+        Returns 400 request if bad payload is sent or it was empty object.
+        """
+        if not request.DATA:
+            error_message = _("No data provided for bookmark")
+            return Response(
+                {
+                    "developer_message": error_message,
+                    "user_message": error_message
+                },
+                status=status.HTTP_400_BAD_REQUEST
+            )
+
+        usage_id = request.DATA.get('usage_id', None)
+        if not usage_id:
+            error_message = _('No usage id provided for bookmark')
+            return Response(
+                {
+                    "developer_message": error_message,
+                    "user_message": error_message
+                },
+                status=status.HTTP_400_BAD_REQUEST
+            )
+
+        try:
+            usage_key = UsageKey.from_string(usage_id)
+
+            # usage_key's course_key may have an empty run property
+            usage_key = usage_key.replace(course_key=modulestore().fill_in_run(usage_key.course_key))
+            course_key = usage_key.course_key
+        except InvalidKeyError:
+            error_message = _(u"invalid usage id '{usage_id}'".format(usage_id=usage_id))
+            return Response(
+                {
+                    "developer_message": error_message,
+                    "user_message": error_message
+                },
+                status=status.HTTP_400_BAD_REQUEST
+            )
+
+        try:
+            module_item = modulestore().get_item(usage_key)
+
+            parent = module_item.get_parent()
+            location_path = []
+            while parent is not None:
+                display_name = parent.display_name if parent.display_name else _("(Unnamed)")
+                location_path.append({"display_name": display_name, "usage_id": unicode(parent.location)})
+                parent = parent.get_parent()
```

Comment:
This logic should be in the model. You can add an optional parameter `block` to `Bookmark.create()` and pass this block to it so that it does not have to be refetched.